pkg/controllers/metrics/node/controller.go

```diff
@@ -0,0 +1,340 @@
+			}
+			return reconcile.Result{}, nil
+		}
+		return reconcile.Result{}, err
+	}
+	// Remove the previous gauge after node labels are updated
+	if labelSlice, ok := c.LabelsMap[req.NamespacedName]; ok {
+		c.deleteGauges(labelSlice)
+	}
+	c.LabelsMap[req.NamespacedName] = []prometheus.Labels{}
+
+	if err := c.updateGauges(ctx, node); err != nil {
+		logging.FromContext(ctx).Debugf("Failed to update gauges: %s", err.Error())
+		return reconcile.Result{}, err
+	}
+
+	return reconcile.Result{}, nil
+}
+
+func (c *Controller) Register(ctx context.Context, m manager.Manager) error {
+	err := controllerruntime.
+		NewControllerManagedBy(m).
+		Named("nodemetrics").
+		For(&v1.Node{}).
+		Watches(
+			// Reconcile all nodes related to a provisioner when it changes.
+			&source.Kind{Type: &v1alpha5.Provisioner{}},
+			handler.EnqueueRequestsFromMapFunc(func(o client.Object) (requests []reconcile.Request) {
+				nodes := &v1.NodeList{}
+				if err := c.KubeClient.List(ctx, nodes, client.MatchingLabels(map[string]string{v1alpha5.ProvisionerNameLabelKey: o.GetName()})); err != nil {
+					logging.FromContext(ctx).Debugf("Failed to list nodes when mapping expiration watch events, %s", err.Error())
+					return requests
+				}
+				for _, node := range nodes.Items {
+					requests = append(requests, reconcile.Request{NamespacedName: types.NamespacedName{Name: node.Name}})
+				}
+				return requests
+			}),
+		).
+		Watches(
+			// Reconcile node when a pod assigned to it changes.
+			&source.Kind{Type: &v1.Pod{}},
+			handler.EnqueueRequestsFromMapFunc(func(o client.Object) (requests []reconcile.Request) {
+				if name := o.(*v1.Pod).Spec.NodeName; name != "" {
+					requests = append(requests, reconcile.Request{NamespacedName: types.NamespacedName{Name: name}})
+				}
+				return requests
+			}),
+		).
+		Complete(c)
+	return err
+}
+
+func (c *Controller) deleteGauges(labelSlice []prometheus.Labels) {
+	for _, labels := range labelSlice {
+		allocatableGaugeVec.Delete(labels)
+		requestsGaugeVec.Delete(labels)
+		limitsGaugeVec.Delete(labels)
+		overheadGaugeVec.Delete(labels)
+	}
+}
+
+// generateLabels creates the labels using the current state of the pod
+func (c *Controller) generateLabels(node *v1.Node, resourceTypeName string) (prometheus.Labels, error) {
+	metricLabels := prometheus.Labels{}
+	metricLabels[resourceType] = resourceTypeName
+	metricLabels[nodeName] = node.GetName()
+	metricLabels[nodeProvisioner] = node.Labels[v1alpha5.ProvisionerNameLabelKey]
+	metricLabels[nodeZone] = node.Labels[v1.LabelTopologyZone]
+	metricLabels[nodeArchitecture] = node.Labels[v1.LabelArchStable]
+	metricLabels[nodeCapacityType] = node.Labels[v1alpha5.LabelCapacityType]
+	metricLabels[nodeInstanceType] = node.Labels[v1.LabelInstanceTypeStable]
+	metricLabels[nodePhase] = string(node.Status.Phase)
+	// Add node labels
+	labels, err := json.Marshal(node.GetLabels())
+	if err != nil {
+		return nil, fmt.Errorf("marshal pod labels: %w", err)
```

Comment:
```suggestion
		return nil, fmt.Errorf("marshalling pod metric labels: %w", err)
```

Reply:
Yeah user won't be able to define prometheus labels. Because when someone defines a label, he/she need to teach the controller how to calculate the values for node/pods.

Reply:
We are no longer returning the node labels since there is no good use case for now.

Reply:
Was this comment in the right place? I wanted to follow convention of using -ing verbs in our `fmt.Errorf()` messages.

Reply:
Yeah user won't be able to define prometheus labels. Because when someone defines a label, he/she need to teach the controller how to calculate the values for node/pods.